### PR TITLE
sort containers by attempt number

### LINKED ISSUE #83050: When the time of a node changes to earlier, the pod may be "CreateContainerError" and can't recover itself
**What happened**:

I have a node "nodeA", I will run a pod "podA" on "nodeA", and "podA" has a container "containerA".
The time now is "timeA".

1. I run "podA" on a "nodeA" at "timeA", so kubelet create a container named "containerA_0", we can use command "docker inspect xxxxxxx" to see the "Created" field of "containerA_0" is "timeA".

2. The time of "nodeA" is changed to "timeB" for some reason. "timeB" is "timeA - 1h".

3. "containerA_0" encounters some errors and become "exited". Kubelet find that and create a container "containerA_1". We can use command "docker inspect xxxxxxxxxx" to see the "Created" fields of "containerA_1" is "timeB".

4. In this condition, when kubelet check the containerStatus of containerA, kubelet will see the "exited" "containerA_0" as the newest continerA , and kubelet will try to  create containerA_1 again and again but failed with error "Conflict ... is already in use by container xxxxxxx".

The reason is:
In `func "getPodContainerStatuses" (pkg/kubelet/kuberuntime/kuberuntime_container.go)`, the containers of podA is sorted by the "Created" field. 
```go
			}
		}
		statuses[i] = cStatus
	}

	sort.Sort(containerStatusByCreated(statuses))
	return statuses, nil
}
```

And `func "computePodActions"(pkg/kubelet/kuberuntime/kuberuntime_manager.go) --> "FindContainerStatusByName"(pkg/kubelet/container/runtime.go)` only pick the first "containerA" of podA's sorted containers, which is "containerA_0".
```go
// FindContainerStatusByName returns container status in the pod status with the given name.
// When there are multiple containers' statuses with the same name, the first match will be returned.
func (podStatus *PodStatus) FindContainerStatusByName(containerName string) *ContainerStatus {
	for _, containerStatus := range podStatus.ContainerStatuses {
		if containerStatus.Name == containerName {
			return containerStatus
		}
	}
	return nil
}
```

Because:
The "Created" field of "containerA_0" is "timeA", 
The "Created" field of "containerA_1" is "timeB",
 "timeB" is "timeA - 1h", so "timeB" < "timeA".
So, kubelet thinks containerA_0 is the newest containerA.
And because containerA_0 is "exited", in order, kubelet will create "containerA_1".
But "containerA_1" is already exist, so it will fail with error "Conflict ... is already in use by container xxxxxxx".

`log:`
```
E0924 09:07:24.225646   30207 remote_runtime.go:200] CreateContainer in sandbox "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" from runtime service failed: rpc error: code = Unknown desc = Error response from daemon: Conflict. The name "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" is already in use by container xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx. You have to remove (or rename) that container to be able to reuse that name.
E0924 09:07:24.225740   30207 kuberuntime_manager.go:779] container start failed: CreateContainerError: Error response from daemon: Conflict. The name "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" is already in use by container xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx. You have to remove (or rename) that container to be able to reuse that name.
E0924 09:07:24.225772   30207 pod_workers.go:190] Error syncing pod xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx ("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx(xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx)"), skipping: failed to "StartContainer" for "composer" with CreateContainerError: "Error response from daemon: Conflict. The name "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" is already in use by container xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx. You have to remove (or rename) that container to be able to reuse that name."
```

5. After failing with the error "Conflict ... is already in use by container xxxxxxx", kubelet will try to remove the existed "containerA_1" to solve the conflict, but "containerA_1" is running, so remove failed.

`func "CreateContainer"(pkg/kubelet/dockershim/docker_container.go) --> "recoverFromCreationConflictIfNeeded" (pkg/kubelet/dockershim/helpers.go)`
```go
// See #33189. If the previous attempt to create a sandbox container name FOO
// failed due to "device or resource busy", it is possible that docker did
// not clean up properly and has inconsistent internal state. Docker would
// not report the existence of FOO, but would complain if user wants to
// create a new container named FOO. To work around this, we parse the error
// message to identify failure caused by naming conflict, and try to remove
// the old container FOO.
// See #40443. Sometimes even removal may fail with "no such container" error.
// In that case we have to create the container with a randomized name.
// TODO(random-liu): Remove this work around after docker 1.11 is deprecated.
// TODO(#33189): Monitor the tests to see if the fix is sufficient.
func recoverFromCreationConflictIfNeeded(client libdocker.Interface, createConfig dockertypes.ContainerCreateConfig, err error) (*dockercontainer.ContainerCreateCreatedBody, error) {
	matches := conflictRE.FindStringSubmatch(err.Error())
	if len(matches) != 2 {
		return nil, err
	}

	id := matches[1]
	klog.Warningf("Unable to create pod sandbox due to conflict. Attempting to remove sandbox %q", id)
	if rmErr := client.RemoveContainer(id, dockertypes.ContainerRemoveOptions{RemoveVolumes: true}); rmErr == nil {
		klog.V(2).Infof("Successfully removed conflicting container %q", id)
		return nil, err
	} else {
		klog.Errorf("Failed to remove the conflicting container %q: %v", id, rmErr)
		// Return if the error is not container not found error.
		if !libdocker.IsContainerNotFoundError(rmErr) {
			return nil, err
		}
	}

	// randomize the name to avoid conflict.
	createConfig.Name = randomizeName(createConfig.Name)
	klog.V(2).Infof("Create the container with randomized name %s", createConfig.Name)
	return client.CreateContainer(createConfig)
}
```

`log:`
```
W0924 09:07:39.224347   30207 helpers.go:286] Unable to create pod sandbox due to conflict. Attempting to remove sandbox "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
E0924 09:07:39.225159   30207 helpers.go:291] Failed to remove the conflicting container "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx": Error response from daemon: You cannot remove a running container xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx. Stop the container before attempting removal or use -f
```

6. In the problem, the pod have containStatus "CreateContainerError".
Even if the time of nodeA returns back to normal later, the problem can't recover itself. I must execute "docker rm -f <containerA_1>" or "docker rm <containerA_0>", then it will recover.
 

**What you expected to happen**:
After the time of nodeA returns back to normal, the problem can recover itself.

**How to reproduce it (as minimally and precisely as possible)**:
1. Run a pod with one container on a node.
2. Use command "date -s xxxxxxx" to change the time of the node to one hour early.
3. Use command "docker kill xxxxxxx" to kill the container of the pod.
4. The problem is now reproduced, and even if we change the time of the node back to nomal, the problem can't recover itself.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):  kubelet version is 1.15.
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### LINKED ISSUE #90298: time-shifting cause kubelet failed to create container
**What happened**:   

time-shifting cause  container created in the feature.
when time shift back to now a days, kubelet will failed to created new container.

```s
Apr 20 20:25:19 master kubelet[3580]: E0420 20:25:19.876407    3580 remote_runtime.go:200] CreateContainer in sandbox "f9dc950f0fdd6aadc8952849622ab6ae8ff6195a73905cdc90dadbf13f80dad2" from
 runtime service failed: rpc error: code = Unknown desc = Error response from daemon: Conflict. The container name "/k8s_echo_echo-58d57fb7c-nrl5j_default_a9365d2a-d8cd-4a9d-948a-d84e064116
84_2" is already in use by container "0eaf56444980451ed2ca3f99bf09a571b61a6023f36519fc35ad3fe6375d5449". You have to remove (or rename) that container to be able to reuse that name.
```

**What you expected to happen**:

kubelet should create container

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**: 
- Kubernetes version (use `kubectl version`): 1.15.1

kubelet sort containers by create timestame which is the root cause of the problem.
sort by `Metadata.Attempt`  sames to solve this problem.

https://github.com/kubernetes/kubernetes/blob/1c62d2c368919de8bba93ea4706632bd2ac8027c/pkg/kubelet/kuberuntime/helpers.go#L52-L62

https://github.com/kubernetes/kubernetes/blob/d4bfcd1fda63aeb984eb0d83f5cc58768e359679/pkg/kubelet/dockershim/naming.go#L58-L79

/sig node

### PR DESCRIPTION
**Pull request by l1b0k**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

kubelet always get Newest container status .
If creation timestamps is not realiable, kubelet will inspect wrong container ,Pod will be in crash state.

**Which issue(s) this PR fixes**:
Fixes #90298
Fixes #83050 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```
/sig node

### CODE REVIEW

File: pkg/kubelet/kuberuntime/helpers.go [hunk 1]
```diff
@@ -42,17 +42,27 @@ func (b containersByID) Swap(i, j int)      { b[i], b[j] = b[j], b[i] }
...
+type podSandboxByAttempt []*runtimeapi.PodSandbox
 
-func (p podSandboxByCreated) Len() int           { return len(p) }
-func (p podSandboxByCreated) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
-func (p podSandboxByCreated) Less(i, j int) bool { return p[i].CreatedAt > p[j].CreatedAt }
+func (p podSandboxByAttempt) Len() int      { return len(p) }
+func (p podSandboxByAttempt) Swap(i, j int) { p[i], p[j] = p[j], p[i] }
+func (p podSandboxByAttempt) Less(i, j int) bool {
+	if p[i].Metadata.Attempt == p[j].Metadata.Attempt {
+		return p[i].CreatedAt > p[j].CreatedAt
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/kuberuntime/helpers.go:51]: should it be reverse?

File: pkg/kubelet/kuberuntime/kuberuntime_gc.go [hunk 2]
```diff
@@ -97,18 +101,28 @@ func (cu containersByEvictUnit) NumEvictUnits() int {
...
+type sandboxByAttempt []sandboxGCInfo
 
-func (a sandboxByCreated) Len() int           { return len(a) }
-func (a sandboxByCreated) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
-func (a sandboxByCreated) Less(i, j int) bool { return a[i].createTime.After(a[j].createTime) }
+func (a sandboxByAttempt) Len() int      { return len(a) }
+func (a sandboxByAttempt) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a sandboxByAttempt) Less(i, j int) bool {
+	if a[i].attempt == a[j].attempt {
+		return a[i].createTime.After(a[j].createTime)
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/kuberuntime/kuberuntime_gc.go:122]: same, it most likely must be opposite condition as this is a `Less` implementation

```suggestion
		return a[i].createTime.Before(a[j].createTime)
```

File: pkg/kubelet/kuberuntime/kuberuntime_gc.go [hunk 3]
```diff
@@ -97,18 +101,28 @@ func (cu containersByEvictUnit) NumEvictUnits() int {
 }
 
 // Newest first.
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/kuberuntime/kuberuntime_gc.go:103]: is this comment still applicable is the sorting is by attempt? Can old container have a very big attempt number?

**Review by SergeyKanzhelev — CHANGES_REQUESTED** [review]

File: pkg/kubelet/kubelet_pods.go [hunk 4]
```diff
@@ -1551,7 +1551,7 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 	}
 
 	// Make the latest container status comes first.
-	sort.Sort(sort.Reverse(kubecontainer.SortContainerStatusesByCreationTime(podStatus.ContainerStatuses)))
+	sort.Sort(sort.Reverse(kubecontainer.SortContainerStatusesByAttempt(podStatus.ContainerStatuses)))
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/kubelet_pods.go:1554]: is that revers Less implementation was intended below, I wonder why it wasn't used here to eliminate the call to `Reverse`

File: pkg/kubelet/kuberuntime/kuberuntime_gc.go [hunk 5]
```diff
@@ -310,6 +326,11 @@ func (cgc *containerGC) evictSandboxes(evictTerminatedPods bool) error {
 		sandboxesByPod[podUID] = append(sandboxesByPod[podUID], sandboxInfo)
 	}
 
+	// Sort the sandboxes by age.
+	for uid := range sandboxesByPod {
```
Comment (l1b0k) [inline pkg/kubelet/kuberuntime/kuberuntime_gc.go:330]: this is deleted in master,i will fix it

File: pkg/kubelet/kubelet_pods.go (hunk 4 above)
Comment (l1b0k) [reply pkg/kubelet/kubelet_pods.go:1554]: this can be simplified

File: pkg/kubelet/kuberuntime/helpers.go (hunk 1 above)
Comment (l1b0k) [reply pkg/kubelet/kuberuntime/helpers.go:51]: i think not.

Comment (SergeyKanzhelev) [reply pkg/kubelet/kuberuntime/helpers.go:51]: Now as the previous definition is fixed and doesn't require the call to `Reverse`, it makes sense.

File: pkg/kubelet/kuberuntime/kuberuntime_gc.go (hunk 3 above)
Comment (l1b0k) [reply pkg/kubelet/kuberuntime/kuberuntime_gc.go:103]: Attempt number is increased when change is detecated.   https://github.com/kubernetes/kubernetes/blob/d140769e4d26f3a175d8a59782ea63595a9f4899/pkg/kubelet/kuberuntime/kuberuntime_container.go#L147-L151     
https://github.com/kubernetes/kubernetes/blob/d140769e4d26f3a175d8a59782ea63595a9f4899/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L442-L445

**Comment by xingfeng1234** [comment]
hi@l1b0k ，now I have met this problem，if you change the time of node，you can reproduce the problem easily. Now I want to solve this problem too。From your conversation with @SergeyKanzhelev ，I have no idea that why the PR is not
merged. Could you interpret that to me？Thank you！

**Comment by l1b0k** [comment]
> 


> hi@l1b0k ，now I have met this problem，if you change the time of node，you can reproduce the problem easily. Now I want to solve this problem too。From your conversation with @SergeyKanzhelev ，I have no idea that why the PR is not
> merged. Could you interpret that to me？Thank you！

It is long time, I think that is a reasonable solution for problem , if this problem is still there there i can re-work on this pr.

**Comment by lilongfeng0902** [comment]
> > 
> 
> > hi@l1b0k ，now I have met this problem，if you change the time of node，you can reproduce the problem easily. Now I want to solve this problem too。From your conversation with @SergeyKanzhelev ，I have no idea that why the PR is not
> > merged. Could you interpret that to me？Thank you！
> 
> It is long time, I think that is a reasonable solution for problem , if this problem is still there there i can re-work on this pr.

Thank for your reply. The account "xingfeng1234" is my another accout, it's me too. ^0^……  I think that the problem is still exists. You can repduce it by the issue [#83050](https://github.com/kubernetes/kubernetes/issues/83050). By the way, I have replace the kubelet binary file(1.20.2) with 1.22.7 version or 1.23.4 version. The same problem is reproduced.  Maybe you could re-work on this pr in the future. Thank you again, if nessory, I can provide more info .

**Comment by l1b0k** [comment]
@lilongfeng0902  This problem still there and i opened a new PR  https://github.com/kubernetes/kubernetes/pull/108474

**Pull request closed without merging**